### PR TITLE
Rearchitecture

### PR DESCRIPTION
**Pull request by ewanmay**

### CODE REVIEW

**Comment by atatmaja** [comment]
let's get staging even with master and then rebase this pr against staging first before I start reviewing. thanks for getting this in tho!

File: src/App.tsx [hunk 1]
```diff
@@ -16,8 +17,41 @@ import httpClient from "./httpClient";
...
+      }
+      else if (pathname === "/Explore") {
+        dispatch({
+          type: 'SELECT_EXPLORE_OR_ANALYZE',
+          payload: true
+        })
+      }
+    }
+
+    const api = new httpClient()
```
Comment (atatmaja) [inline src/App.tsx:40]: Make comment that this will be updated to use new endpoint to get filter options

File: src/App.tsx [hunk 2]
```diff
@@ -26,32 +60,34 @@ function App() {
...
       });
     }
+
     const getAirtableRecords = async () => {
-      const response = await api.getAirtableRecords()
-      dispatch({
-        type: 'GET_AIRTABLE_RECORDS',
-        payload: response
-      });
+      if (dataPageState.routingOccurred) {
```
Comment (atatmaja) [inline src/App.tsx:65]: Document this weirdness somewhere so we can come up with ideas on how to make this better

File: src/components/charts/Charts.tsx [hunk 3]
```diff
@@ -17,34 +17,34 @@ export default function Charts() {
...
-  const { filtered_records, filters, showAnalyzePopup } = state;
+  const { filters, showAnalyzePopup } = state;
   const initState = [] as AggregatedRecord[];
   const [showModal, toggleModal] = useState(true);
   const [records, setRecords] = useState(initState);
 
   const yAxisOptions = [
     { key: 'Geographies', text: Translate('Geographies'), value: AggregationFactor.country },
-    { key: 'Population', text: Translate('Population'), value: AggregationFactor.population_group },
+    // { key: 'Population', text: Translate('Population'), value: AggregationFactor.population_group },
```
Comment (atatmaja) [inline src/components/charts/Charts.tsx:27]: Can we make a backend ticket to make sure that these fixes get done. I think maybe @simonarocco can take it cause it seems to be related to meta analysis. Will add more details in the tick

File: src/components/charts/Charts.tsx [hunk 4]
```diff
@@ -17,34 +17,34 @@ export default function Charts() {
...
 
   useEffect(() => {
-    if (filtered_records.length > 0) {
-      const updateCharts = async () => {
-        const api = new httpClient();
-        const reAggregatedRecords = await api.postMetaAnalysis(state.filtered_records, yAxisSelection);
-        const chartData = _.sortBy(reAggregatedRecords, 'seroprevalence').reverse();
-        setRecords(chartData);
-      }
+    const updateCharts = async () => {
```
Comment (atatmaja) [inline src/components/charts/Charts.tsx:38]: Maybe we can cut down on the number of api calls we make here.

Consider the following scenario:
1. Go to analyze and load the chart
2. Go to data
3. Go back to analyze --> here shouldn't have to recall api as cached data should suffice

File: src/components/pages/Dashboard/Analyze.tsx [hunk 5]
```diff
@@ -11,20 +11,18 @@ import Charts from "../../charts/Charts";
...
-      const updateCountryPrevalence = async () => {
-        const api = new httpClient();
-        const estimateGradePrevalences = await api.getEstimateGrades(state.filtered_records);
-        dispatch({
-          type: 'UPDATE_ESTIMATE_PREVALENCES',
-          payload: estimateGradePrevalences
-        });
-      } 
-      updateCountryPrevalence();
+    const updateCountryPrevalence = async () => {
```
Comment (atatmaja) [inline src/components/pages/Dashboard/Analyze.tsx:16]: Figure out if we still need this lol, or rather, if it is best placed in analyze.tsx

File: src/components/pages/Dashboard/test.json [hunk 6]
```diff
@@ -0,0 +1,1082 @@
+[
```
Comment (atatmaja) [inline src/components/pages/Dashboard/test.json:1]: @ewanmay plz delete haha

File: src/components/shared/StudiesTable.tsx [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+        else {
+            setTotalPages(Math.ceil(props.dataRecords.length / pageLength));
+        }
+    }, [activePage, column, direction, isMobileDevice, pageLength, props.dataRecords])
+
+    const getPopulation = (sex: string | null, age: string[] | null, population_group: string[] | null) => {
+        if (!population_group) {
+            return "Not Reported";
+        }
+        const displaySex = sex && sex !== "All" && sex !== "Unspecified";
```
Comment (atatmaja) [inline src/components/shared/StudiesTable.tsx:53]: TODO: figure out if we still need to make all these edge case checks

File: src/components/sidebar/left-sidebar/AnalysisMethods.tsx [hunk 8]
```diff
@@ -1,9 +1,8 @@
 import React from "react";
-import { MIN_DENOMINATOR } from "../../../metaAnalysis";
 import InformationText from "../../shared/InformationText";
 import './AnalysisMethods.css';
 
-
+const MIN_DENOMINATOR = 100
```
Comment (atatmaja) [inline src/components/sidebar/left-sidebar/AnalysisMethods.tsx:5]: do we still use this? if not get rid plz

File: src/components/sidebar/left-sidebar/TotalStats.tsx [hunk 9]
```diff
@@ -1,30 +1,28 @@
...
-        const api = new httpClient();
-        const results = await api.postMetaAnalysisAll(state.filtered_records);
-        setNumCountries(results.countries);
-        setSeroprevalence(results.seroprevalence);
-        setN(results.n);
-      } 
-      setNumStudies(state.filtered_records.length);
+    const updateCountryPrevalence = async () => {
+      const api = new httpClient();
+      const filters = state.dataPageState.exploreIsOpen ? state.exploreFilters : state.analyzeFilters
```
Comment (atatmaja) [inline src/components/sidebar/left-sidebar/TotalStats.tsx:16]: should not need to check exploreIsOpen here

validate this tho plz

File: src/context.tsx [hunk 10]
```diff
@@ -56,107 +56,25 @@ const initial_filters: Filters = getDefaultFilters();
...
   estimate_grade_prevalences: [],
-  filters: initial_filters,
-  filter_options: getEmptyFilters(),
-  all_filter_options: getEmptyFilters(),
-  data_page_state: {
-    mapOpen: true
+  exploreFilters: { ...initial_filters, estimate_grade: new Set(), source_type: new Set() },
+  analyzeFilters: initial_filters,
+  filterOptions: getEmptyFilters(),
+  allFilterOptions: getEmptyFilters(),
```
Comment (atatmaja) [inline src/context.tsx:64]: don't think we need this anymore

File: src/context.tsx [hunk 11]
```diff
@@ -56,107 +56,25 @@ const initial_filters: Filters = getDefaultFilters();
...
-    // the user will never get 0 records back if they choose more filter options for a filter
-    // that they've already selected an option for
-    if (filters[filter as FilterType].size > 0) {
-      options[filter as FilterType] = all_filter_options[filter as FilterType];
-    }
-  }
-  return options;
-}
-
 function getFilterOptions(records: AirtableRecord[]) {
```
Comment (atatmaja) [inline src/context.tsx:78]: Again comment a TODO here to delete this function when we integrate the new backend endpoint

File: src/httpClient.ts [hunk 12]
```diff
@@ -1,6 +1,17 @@
-import { AirtableRecord, AggregatedRecord, AggregationFactor, AlternateAggregatedRecord } from "./types"
+import { AggregatedRecord, AggregationFactor, AirtableRecord, Filters, FilterType } from "./types";
+import { formatISO, add } from 'date-fns'
 
 export default class httpClient {
+    formatDates(dates: Array<Date> | null) {
```
Comment (atatmaja) [inline src/httpClient.ts:5]: suggestion: move to utils

File: src/httpClient.ts [hunk 13]
```diff
@@ -41,79 +52,130 @@ export default class httpClient {
...
+            const filter = Array.from(filters[o as FilterType]);
+            if (o !== 'publish_date') {
+                reqBodyFilters[o] = filter as string[]
+            }
+        });
 
-    async getAirtableRecords() {
-        const response = await this.httpGet('/airtable_scraper/records')
-        if (!response || !("records" in response) || !("updated_at" in response)) {
+        const date = filters['publish_date'] as Array<Date>
```
Comment (atatmaja) [inline src/httpClient.ts:110]: I think this should be called "sampling_end_date" instead

File: src/httpClient.ts [hunk 14]
```diff
@@ -41,79 +52,130 @@ export default class httpClient {
...
+            columns: [],
+        }
+        const response = await this.httpPost('/data_provider/records', reqBody)
+        if (!response) {
             return [];
         }
-        const airtable_records = response.records!.map((item: Record<string, any>) => {
+        const filtered_records = response.map((item: Record<string, any>) => {
             // Convert response to AirtableRecord type
             const record: AirtableRecord = {
```
Comment (atatmaja) [inline src/httpClient.ts:128]: suggestion: have some sort of function to return this big airtable record object

File: src/httpClient.ts [hunk 15]
```diff
@@ -41,79 +52,130 @@ export default class httpClient {
...
+        Object.keys(filters).forEach((o: string) => {
+            const filter = Array.from(filters[o as FilterType]);
+            reqBodyFilters[o] = filter as string[]
         })
-        const req_body = {
-            records: formattedRecords
+        delete reqBodyFilters['publish_date']
+        const date = filters['publish_date']
+        const endDate = date[1] ? formatISO(date[1] as Date) : formatISO(new Date());
+        const startDate = date[0] ? formatISO(date[0] as Date) : formatISO(add(new Date(endDate), { years: -2 }));
```
Comment (atatmaja) [inline src/httpClient.ts:171]: use date parsing utility function here!

Comment (atatmaja) [reply src/httpClient.ts:171]: On a second note, would be better for us to just be able to not provide a startDate to the backend if it's not specified. 

TODO: Validate if this functionality exists, if not then make ticket to do it
@atatmaja

File: src/httpClient.ts [hunk 16]
```diff
@@ -150,73 +212,43 @@ export default class httpClient {
         return formattedResponse;
     }
 
-    async postMetaAnalysis(records: AirtableRecord[], aggregation_variable: AggregationFactor, meta_analysis_technique: string = 'fixed', meta_analysis_transformation: string = 'double_arcsin_precise') {
+    async postMetaAnalysis(filters: Filters,
+        aggregation_variable: AggregationFactor,
+        meta_analysis_technique: string = 'fixed',
+        meta_analysis_transformation: string = 'double_arcsin_precise') {
         // Note: while the rest of the aggregation variables can stay consistent with frontend nomenclature
```
Comment (atatmaja) [inline src/httpClient.ts:219]: incomplete sentence plz take gr 11 english again

File: src/tests/system.test.ts [hunk 17]
```diff
@@ -0,0 +1,5 @@
+import { testRecords } from './testData'	
+
+test('placeholder test', () => {
```
Comment (atatmaja) [inline src/tests/system.test.ts:3]: wow so robust!!!!!!

Comment (atatmaja) [reply src/tests/system.test.ts:3]: TODO make real tests

**Review by atatmaja — COMMENTED** [review]
Few minor ish suggestions. For all of the "TODO" items, please make comments and tickets. Good work, starting to see the light at the end of the tunnel!!!!!!!!

File: src/components/sidebar/left-sidebar/AnalysisMethods.tsx (hunk 8 above)
Comment (ewanmay) [reply src/components/sidebar/left-sidebar/AnalysisMethods.tsx:5]: we do still use it apparently

**Pull request merged**